Userland/Libraries/LibJS/AST.h

```diff
@@ -97,6 +97,45 @@ class ASTNode : public RefCounted<ASTNode> {
     u32 m_end_offset { 0 };
 };
 
+// This is a helper class that packs an array of T after the AST node, all in the same allocation.
+template<typename Derived, typename Base, typename T>
+class ASTNodeWithTailArray : public Base {
+public:
+    template<typename... Args>
+    static NonnullRefPtr<Derived> create(size_t tail_size, SourceRange source_range, Args&&... args)
+    {
+        static_assert(alignof(Derived) % alignof(T) == 0, "need padding for tail array");
```

Comment:
I would suggest also adding a check for `std::is_final_v<Derived>`, otherwise the tail data might override class members of a further derived class.

Reply:
That would break with `NewExpression` already inheriting from `CallExpression` though. I've instead added a `static_assert` about `sizeof(NewExpression) == sizeof(CallExpression)` in a subsequent commit. It's not ideal, but it's something.